src/components/app_catalog/app_list/list_inner.js

```diff
@@ -0,0 +1,178 @@
+import { replace } from 'connected-react-router';
+import AppListItems from './list_items';
+import AppListSearch from './list_search';
+import lunr from 'lunr';
+import PropTypes from 'prop-types';
+import React from 'react';
+
+class AppListInner extends React.Component {
+  state = {
+    filters: [],
+    searchQuery: null,
+    iconErrors: {},
+  };
+
+  // Contains refs to all the app-container divs in dom so that we can
+  // scroll to them if needed.
+  appRefs = {};
+
+  index = null;
+
+  constructor(props) {
+    super(props);
+
+    this.index = lunr(function() {
+      this.ref('name');
+      this.field('name');
+      this.field('description');
+      this.field('keywords');
+
+      const apps = Object.values(props.catalog.apps).map(appVersions => {
+        return appVersions.sort((a, b) => {
+          return new Date(b.created) - new Date(a.created);
+        })[0];
+      });
+
+      for (const app of apps) this.add(app);
+    });
+  }
+
+  static getDerivedStateFromProps(nextProps, prevState) {
+    const searchParams = new URLSearchParams(nextProps.location.search);
+    const currentSearchQuery = searchParams.get('q') || '';
+
+    if (prevState.searchQuery === null) {
+      return {
+        searchQuery: currentSearchQuery,
+      };
+    }
+
+    return null;
+  }
```

Comment:
This look like a code smell for me, ¿Could we do this in `ComponentDidUpdate` instead?

Reply:
No, this is the recommended way of setting the state based on a props change. Here I'm using it to set the initial state based on the props.

Reply:
Well, `componentDidUpdate` won't rerender anything if you compare props inside it and just triggers the ones you want.

As for derived state, sorry but I don't get the point of `getDerivedStateFromProps` here... 
Now that I am rereading it, with the code you have inside `getDerivedStateFromProps` and initializing `this.state.searchQuery` with `null` what this would do is just set an empty string in `searchQuery` when component mounts, right? Why don't initialize `searchQuery`with `''` instead? 

TBH I never use `getDerivedStateFromProps` so I might be missing something.

Thanks!

Reply:
`getDerivedStateFromProps` merges the update triggered by changing props and the update triggered by changing state. `componentDidUpdate` doesn't re-render anything, unless you change the state in it. (which just means you're triggering another update after an update)

As said before, here I'm using it to set the initial state based on the props. Setting the initial state based on props in the constructor is an anti-pattern.

I'm setting `searchQuery`'s initial value to `null`, so I could diferentiate the initial value of it, to the value that it would have if the search input would be empty. 

Having `if (prevState.searchQuery === '') {` would mean that a state change would be triggered every time the props change.

Reply:
> `getDerivedStateFromProps` merges the update triggered by changing props and the update triggered by changing state. `componentDidUpdate` doesn't re-render anything, unless you change the state in it. (which just means you're triggering another update after an update)

Ok if this is what you need here.

 
> As said before, here I'm using it to set the initial state based on the props. Setting the initial state based on props in the constructor is an anti-pattern.

Yeah, that's the mantra, but it's an oversimplification, basically this is not true. We do it all the time. It is an antipattern as long as you expect this to trigger rerenders. Otherwise it is perfectly fine.

> I'm setting `searchQuery`'s initial value to `null`, so I could diferentiate the initial value of it, to the value that it would have if the search input would be empty.
> 
> Having `if (prevState.searchQuery === '') {` would mean that a state change would be triggered every time the props change.

Don't agree, but anyway, that's not the point I was willing to raise. I am just asking what problem does this code actually solve, because as far as I understand this just runs when component is mounted, and when this happens it will set `searchQuery` always to `' '` because it is initialized as `null`. So the question is ¿Does this code run in other situations? 

Just asking because I want understand it, that's all.

Reply:
> > As said before, here I'm using it to set the initial state based on the props. Setting the initial state based on props in the constructor is an anti-pattern.
> 
> Yeah, that's the mantra, but it's an oversimplification, basically this is not true. We do it all the time. It is an antipattern as long as you expect this to trigger rerenders. Otherwise it is perfectly fine.

It's actually always an anti-pattern. The reason is because of React Fiber, the component might be instantiated a while after the component has been mounted, and if that happens, the props might not be there yet. That will become an even bigger problem later when React makes rendering asynchronous.

> > I'm setting `searchQuery`'s initial value to `null`, so I could diferentiate the initial value of it, to the value that it would have if the search input would be empty.
> > Having `if (prevState.searchQuery === '') {` would mean that a state change would be triggered every time the props change.
> 
> Don't agree, but anyway, that's not the point I was willing to raise. I am just asking what problem does this code actually solve, because as far as I understand this just runs when component is mounted, and when this happens it will set `searchQuery` always to `' '` because it is initialized as `null`. So the question is ¿Does this code run in other situations?

This does not set `searchQuery` to always be `''`. If there is a `q` parameter in the URL. This sets the `searchQuery` value to the one in the URL. So for example if you send the URL (`some_URL?q=someApp`) to someone, `searchQuery` would be set to `someApp` for the initial render, and the browser would scroll to that app.

Reply:
We discussed about that, don't wanna make this longer than what it actually is, just could you please take a look at here? https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops and here: https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization and decide if this fits in some of the cases where we don't need `getDerivedStateFromProps`.

I'm ok with whatever you decide.